File: R/nat.fastcore/tests/testthat/test-cmtk.R
# CMTK transforms, checked against CMTK's own `streamxform` output.
#
# The ground truth in `streamxform_golden.csv` was produced by the CMTK binary itself, so
# these are not self-consistency checks: they pin us to CMTK's behaviour, including its habit
# of *failing* on points it cannot transform (which we return as NaN).

reg_path <- system.file("extdata", "JFRC2_FCWB.list", package = "nat.fastcore")
tiny_path <- system.file("extdata", "tiny_warp.list", package = "nat.fastcore")

golden <- local({
  g <- utils::read.csv(system.file("extdata", "streamxform_golden.csv",
                                   package = "nat.fastcore"))
  split(as.matrix(g[, c("x", "y", "z")]), g$case)[unique(g$case)] |>
    lapply(matrix, ncol = 3)
})

# Measured margin is ~5e-7 (streamxform's own print precision). If a change lands at 1e-5,
# something is subtly wrong -- do not loosen this.
ATOL <- 1e-4

reg <- cmtk_read(reg_path)
pts <- golden$input

test_that("affine and warp match streamxform, forwards", {
  expect_equal(cmtk_xform(reg, pts, transform = "affine"), golden$affine_forward,
               tolerance = ATOL, ignore_attr = TRUE)
  expect_equal(cmtk_xform(reg, pts), golden$warp_forward,
               tolerance = ATOL, ignore_attr = TRUE)
})

test_that("affine and warp match streamxform, backwards", {
  expect_equal(cmtk_xform_inv(reg, pts, transform = "affine"), golden$affine_inverse,
               tolerance = ATOL, ignore_attr = TRUE)
  expect_equal(cmtk_xform_inv(reg, pts), golden$warp_inverse,
               tolerance = ATOL, ignore_attr = TRUE)
})

test_that("points CMTK reports as FAILED come back NaN", {
  # streamxform fails on 2 of the 5 sample points; returning a plausible-looking number
  # instead would silently disagree with every other CMTK-based tool.
  out <- cmtk_xform_inv(reg, pts)
  expect_true(all(is.nan(out[c(1, 5), ])))
  expect_true(all(is.finite(out[2:4, ])))
})

test_that("round-trips", {
  p <- pts[2:4, ]
  expect_equal(cmtk_xform_inv(reg, cmtk_xform(reg, p)), p, tolerance = 1e-4)
  expect_equal(
    cmtk_xform_inv(reg, cmtk_xform(reg, p, transform = "affine"), transform = "affine"),
    p, tolerance = 1e-10
  )
})

test_that("out-of-domain points are NaN by default", {
  far <- rbind(c(-5000, -5000, -5000), c(100, 100, 20))

  strict <- cmtk_xform(reg, far)
  expect_true(all(is.nan(strict[1, ])))
  expect_true(all(is.finite(strict[2, ])))

  # ...unless you opt into extrapolation, which makes you disagree with CMTK
  expect_true(all(is.finite(cmtk_xform(reg, far, allow_extrapolation = TRUE))))

  # ...or fall back to the affine
  filled <- cmtk_xform(reg, far, fallback_to_affine = TRUE)
  expect_equal(filled[1, ], cmtk_xform(reg, far, transform = "affine")[1, ])
})

test_that("the affine fallback survives a backwards hop", {
  # Regression: `fallback_to_affine` used to be a silent no-op in the inverse direction,
  # which is the one navis walks about half of its bridging edges in.
  far <- rbind(c(-5000, -5000, -5000), c(100, 100, 20))

  expect_true(all(is.nan(cmtk_xform(reg, far, invert = TRUE)[1, ])))

  filled <- cmtk_xform(reg, far, invert = TRUE, fallback_to_affine = TRUE)
  expect_true(all(is.finite(filled)))
  # ...and it falls back to the *inverse* affine, not the forward one.
  expect_equal(filled[1, ], cmtk_xform(reg, far, invert = TRUE, transform = "affine")[1, ])
  expect_gt(max(abs(filled[1, ] - cmtk_xform(reg, far, transform = "affine")[1, ])), 1)

  # The same hole from the other side: `cmtk_xform_inv` could not fall back at all.
  expect_true(all(is.nan(cmtk_xform_inv(reg, far)[1, ])))
  filled_inv <- cmtk_xform_inv(reg, far, fallback_to_affine = TRUE)
  expect_equal(filled_inv[1, ], cmtk_xform_inv(reg, far, transform = "affine")[1, ])
})

test_that("the affine fallback covers the whole chain, as nat/navis do", {
  # `streamxform --affine-only` composes the affine of every registration in the list, and
  # navis re-runs failed rows through it from the ORIGINAL point -- so a point that survives
  # hop 1 and fails hop 2 loses its hop-1 warp. "hop" keeps it, but departs from CMTK.
  chain <- cmtk_read(c(reg_path, reg_path))
  p <- rbind(c(5, 179, 38)) # inside the domain for hop 1, outside it for hop 2
  expect_true(any(is.nan(cmtk_xform(chain, p))))

  by_chain <- cmtk_xform(chain, p, fallback_to_affine = TRUE)
  by_hop <- cmtk_xform(chain, p, fallback_to_affine = "hop")
  expect_true(all(is.finite(by_chain)))
  expect_true(all(is.finite(by_hop)))

  # TRUE == "chain" == the whole chain, affine-only, from the original point
  expect_equal(by_chain, cmtk_xform(chain, p, transform = "affine"))
  expect_equal(by_chain, cmtk_xform(chain, p, fallback_to_affine = "chain"))
  # ...and "hop" lands elsewhere, having kept the hop-1 warp
  expect_gt(max(abs(by_chain - by_hop)), 1)

  expect_error(cmtk_xform(reg, pts, fallback_to_affine = "bogus"), "chain")
})

test_that("input coercion", {
  # a bare length-3 vector, and a data frame, both work
  expect_equal(cmtk_xform(reg, c(50, 50, 50)), cmtk_xform(reg, rbind(c(50, 50, 50))))
  df <- as.data.frame(pts)
  expect_equal(cmtk_xform(reg, df), cmtk_xform(reg, pts))
})

test_that("chains and the invert flag", {
  chain <- cmtk_read(c(reg_path, reg_path))
  expect_length(cmtk_versions(chain), 2)
  p <- pts[2:4, ]
  expect_equal(cmtk_xform(chain, p), cmtk_xform(reg, cmtk_xform(reg, p)))

  # traversing a registration backwards is its inverse -- from the same parse
  expect_equal(cmtk_xform(reg, cmtk_xform(reg, p), invert = TRUE), p, tolerance = 1e-4)

  # and inverting the whole chain undoes both hops
  expect_equal(cmtk_xform_inv(chain, cmtk_xform(chain, p)), p, tolerance = 1e-4)

  # `invert` is per hop, so a chain can be walked in mixed directions -- which no whole-chain
  # inversion can express. Affine-only, so nothing falls out of a domain and becomes NaN.
  mixed <- cmtk_xform(chain, p, transform = "affine", invert = c(FALSE, TRUE))
  expect_true(all(is.finite(mixed)))
  expect_equal(mixed, p, tolerance = 1e-4) # A then A-inverse, over the same registration
})

test_that("n_cores does not change the answer", {
  expect_equal(cmtk_xform_inv(reg, pts, n_cores = 1), cmtk_xform_inv(reg, pts))
})

test_that("plain and gzipped registrations both load", {
  # JFRC2 is gzipped, the tiny fixture is plain text; gzip is sniffed from magic bytes.
  expect_equal(cmtk_versions(reg), "1.1")
  expect_equal(cmtk_versions(cmtk_read(tiny_path)), "2.4")
  expect_equal(as.vector(cmtk_dims(reg)), c(59, 27, 11))
  expect_equal(as.vector(cmtk_dims(cmtk_read(tiny_path))), c(5, 5, 5))
})

test_that("properties", {
  expect_equal(dim(cmtk_affine(reg)), c(4L, 4L))
  expect_equal(as.vector(cmtk_affine(reg)[4, ]), c(0, 0, 0, 1))
  expect_equal(as.vector(cmtk_spacing(reg)), c(11.3642, 13.2453, 16.8741), tolerance = 1e-4)
  expect_equal(as.vector(cmtk_domain(reg)), c(636.396, 317.887, 134.9931), tolerance = 1e-4)
  expect_output(print(reg), "cmtk_registration")
})

test_that("errors are informative", {
  expect_error(cmtk_read("/nonexistent/nope.list"), "no CMTK registration")
  expect_error(cmtk_read(tempdir()), "neither")
  expect_error(cmtk_xform(reg, pts, invert = c(TRUE, FALSE)), "one flag per registration")
  expect_error(cmtk_xform(reg, matrix(1, 2, 2)), "\\(N, 3\\)")
  expect_error(cmtk_xform(reg, pts, transform = "bogus"))
  expect_error(cmtk_xform(42, pts), "cmtk_registration")
  expect_error(cmtk_xform_inv(reg, pts, initial_guess = pts[1:2, ]),
               "one point per input point")
})

File: py/tests/test_cmtk.py
"""CMTK transforms: the Rust kernels vs CMTK's own `streamxform`.

The ground truth in `streamxform_golden.csv` was produced by the CMTK binary itself, so
these are not self-consistency checks - they pin us to CMTK's behaviour, including its habit
of *failing* on points whose inverse does not converge (which we return as NaN).
"""

import pickle

import numpy as np
import pytest

import navis_fastcore as fastcore

# `atol=1e-4` is the tolerance of the reference implementation's own golden tests. Our
# measured margin is ~4e-7, i.e. three orders of magnitude of headroom. If a change pushes a
# result to 1e-5, something is subtly wrong - do not loosen this.
ATOL = 1e-4
RTOL = 1e-5


@pytest.fixture(scope="module")
def reg(cmtk_dir):
    return fastcore.CmtkRegistration(cmtk_dir)


# ---------------------------------------------------------------------------
# Against streamxform
# ---------------------------------------------------------------------------


@pytest.mark.parametrize(
    "case,kwargs",
    [
        ("affine_forward", dict(transform="affine")),
        ("warp_forward", dict(transform="warp")),
    ],
)
def test_forward_matches_streamxform(reg, golden, case, kwargs):
    out = reg.xform(golden["input"], **kwargs)
    np.testing.assert_allclose(out, golden[case], atol=ATOL, rtol=RTOL, equal_nan=True)


@pytest.mark.parametrize(
    "case,kwargs",
    [
        ("affine_inverse", dict(transform="affine")),
        ("warp_inverse", dict(transform="warp")),
    ],
)
def test_inverse_matches_streamxform(reg, golden, case, kwargs):
    out = reg.xform_inv(golden["input"], **kwargs)
    np.testing.assert_allclose(out, golden[case], atol=ATOL, rtol=RTOL, equal_nan=True)


def test_warp_inverse_reproduces_streamxform_failures(reg, golden):
    """CMTK reports two of the five sample points as FAILED; we must return NaN there.

    Getting a finite answer for these would be *worse* than useless: it would silently
    disagree with every other CMTK-based tool.
    """
    out = reg.xform_inv(golden["input"])
    assert np.isnan(out[0]).all()
    assert np.isnan(out[4]).all()
    assert np.isfinite(out[[1, 2, 3]]).all()


# ---------------------------------------------------------------------------
# Round-trips
# ---------------------------------------------------------------------------


def test_warp_roundtrip(reg):
    pts = np.array(
        [[50.0, 50.0, 50.0], [100.0, 100.0, 20.0], [250.0, 150.0, 60.0], [300.0, 120.0, 70.0]]
    )
    back = reg.xform_inv(reg.xform(pts))
    np.testing.assert_allclose(back, pts, atol=1e-4)


def test_affine_roundtrip_is_exact(reg):
    pts = np.array([[0.0, 0.0, 0.0], [123.0, 45.0, 67.0], [500.0, 250.0, 100.0]])
    back = reg.xform_inv(reg.xform(pts, transform="affine"), transform="affine")
    np.testing.assert_allclose(back, pts, atol=1e-10)


# ---------------------------------------------------------------------------
# Input handling
# ---------------------------------------------------------------------------


def test_single_point_in_single_point_out(reg):
    single = reg.xform(np.array([50.0, 50.0, 50.0]))
    assert single.shape == (3,)
    batch = reg.xform(np.array([[50.0, 50.0, 50.0]]))
    assert batch.shape == (1, 3)
    np.testing.assert_allclose(single, batch[0])


def test_int_input_is_coerced(reg):
    out = reg.xform(np.array([[50, 50, 50]], dtype=np.int64))
    np.testing.assert_allclose(out, reg.xform(np.array([[50.0, 50.0, 50.0]])))


def test_list_input_works(reg):
    np.testing.assert_allclose(
        reg.xform([[50.0, 50.0, 50.0]]), reg.xform(np.array([[50.0, 50.0, 50.0]]))
    )


def test_bad_shape_raises(reg):
    with pytest.raises(ValueError, match=r"\(N, 3\)"):
        reg.xform(np.zeros((4, 2)))


def test_bad_transform_raises(reg):
    with pytest.raises(ValueError, match="warp.*affine"):
        reg.xform(np.zeros((1, 3)), transform="bogus")


def test_missing_path_raises():
    with pytest.raises(ValueError, match="no such file or directory"):
        fastcore.CmtkRegistration("/nonexistent/nope.list")


def test_initial_guess_length_mismatch_raises(reg):
    with pytest.raises(ValueError, match="one point per input point"):
        reg.xform_inv(np.zeros((3, 3)), initial_guess=np.zeros((2, 3)))


# ---------------------------------------------------------------------------
# Behaviour flags
# ---------------------------------------------------------------------------


def test_out_of_domain_is_nan_by_default_and_fallback_fills_it(reg):
    """CMTK reports points outside the domain box as FAILED; we return NaN, by default.

    Returning a clamped-boundary extrapolation instead (which is what the reference
    implementation does by default) would silently disagree with every CMTK-based tool.
    """
    pts = np.array([[-5000.0, -5000.0, -5000.0], [100.0, 100.0, 20.0]])

    strict = reg.xform(pts)
    assert np.isnan(strict[0]).all()
    assert np.isfinite(strict[1]).all()

    filled = reg.xform(pts, fallback_to_affine=True)
    np.testing.assert_allclose(filled[0], reg.xform(pts, transform="affine")[0], atol=1e-12)
    # The in-domain point is untouched by the fallback.
    np.testing.assert_allclose(filled[1], strict[1])


def test_fallback_to_affine_survives_an_inverted_registration(cmtk_dir):
    """Regression: `fallback_to_affine` used to be a silent no-op when `invert=True`.

    This is the direction navis walks about half of its bridging edges in - it passes
    `affine_fallback=True` by default through `TransformSequence` - so an inverted edge
    silently dropped every point the binary path would have rescued.
    """
    reg = fastcore.CmtkRegistration(cmtk_dir)
    pts = np.array([[-5000.0, -5000.0, -5000.0], [100.0, 100.0, 20.0]])

    strict = reg.xform(pts, invert=True)
    assert np.isnan(strict[0]).all()
    assert np.isfinite(strict[1]).all()

    filled = reg.xform(pts, invert=True, fallback_to_affine=True)
    assert np.isfinite(filled).all()

    # ...and it falls back to the *inverse* affine, not the forward one. A wrong-direction
    # fallback would be worse than a NaN: a plausible number pointing into the wrong space.
    inv_affine = reg.xform(pts, invert=True, transform="affine")
    fwd_affine = reg.xform(pts, transform="affine")
    np.testing.assert_allclose(filled[0], inv_affine[0], atol=1e-12)
    assert np.abs(inv_affine[0] - fwd_affine[0]).max() > 1.0, "else this proves nothing"


def test_fallback_to_affine_survives_xform_inv(reg):
    """The same hole from the other side: `xform_inv` could not fall back at all."""
    pts = np.array([[-5000.0, -5000.0, -5000.0], [100.0, 100.0, 20.0]])

    assert np.isnan(reg.xform_inv(pts)[0]).all()

    filled = reg.xform_inv(pts, fallback_to_affine=True)
    np.testing.assert_allclose(
        filled[0], reg.xform_inv(pts, transform="affine")[0], atol=1e-12
    )


def test_chain_and_hop_fallbacks_differ_on_a_chain(cmtk_dir):
    """`fallback_to_affine` falls back over the *whole chain*, as nat/navis do.

    Verified against the binary: `streamxform --affine-only` composes the affine of every
    registration in the list, and navis re-runs the failed rows through it from the ORIGINAL
    point. So a point that survives hop 1 and fails hop 2 loses its hop-1 warp.

    `"hop"` keeps that warp and swaps the affine in only where it actually failed. Better,
    probably - but a silent departure, so it has to be asked for.
    """
    chain = fastcore.CmtkRegistration([cmtk_dir, cmtk_dir])
    pts = np.array([[5.0, 179.0, 38.0]])  # inside the domain for hop 1, outside it for hop 2

    assert np.isnan(chain.xform(pts)).any(), "this point must fail, or the test proves nothing"

    by_chain = chain.xform(pts, fallback_to_affine=True)
    by_hop = chain.xform(pts, fallback_to_affine="hop")
    assert np.isfinite(by_chain).all() and np.isfinite(by_hop).all()

    # True == "chain" == the whole chain, affine-only, from the original point
    np.testing.assert_allclose(by_chain, chain.xform(pts, transform="affine"), atol=1e-12)
    np.testing.assert_allclose(by_chain, chain.xform(pts, fallback_to_affine="chain"), atol=1e-12)

    # ...and "hop" lands somewhere else entirely, because it kept the hop-1 warp
    assert np.abs(by_chain - by_hop).max() > 1.0


def test_on_a_single_registration_chain_and_hop_agree(reg):
    """The two only diverge once there is more than one hop to disagree about."""
    pts = np.array([[-5000.0, -5000.0, -5000.0]])
    np.testing.assert_allclose(
        reg.xform(pts, fallback_to_affine="chain"),
        reg.xform(pts, fallback_to_affine="hop"),
        atol=1e-12,
    )


def test_bad_fallback_value_raises(reg):
    with pytest.raises(ValueError, match="chain"):
        reg.xform(np.zeros((1, 3)), fallback_to_affine="bogus")


def test_extrapolation_can_be_opted_into(reg):
    out = reg.xform(np.array([[-5000.0, -5000.0, -5000.0]]), allow_extrapolation=True)
    assert np.isfinite(out).all()


def test_domain_is_the_world_box_not_the_lattice_extent(reg):
    """CMTK's domain is `0 <= p <= domain`, NOT the control-point lattice's extent.

    The lattice is padded one cell outside the domain (origin == -spacing), so a point with
    lattice coordinate u just below 1 sits inside the lattice box but *outside* the world
    domain box. CMTK fails it. The reference implementation accepts it.
    """
    domain = reg.spacing[0] * (reg.dims[0] - 3)
    origin = -reg.spacing[0]

    # u_x = 0.5 -> inside the lattice box [0, dims-3), but world x < 0 -> outside the domain.
    p = np.array([[origin[0] + 0.5 * reg.spacing[0][0], 100.0, 50.0]])
    assert p[0, 0] < 0
    assert np.isnan(reg.xform(p)).all()

    # The domain corners themselves are in.
    assert np.isfinite(reg.xform(np.array([[0.0, 0.0, 0.0]]))).all()
    assert np.isfinite(reg.xform(np.array([domain - 1e-6]))).all()


def test_clamp_to_domain_off_recovers_out_of_domain_preimages(reg, golden):
    """The two points CMTK calls FAILED do have preimages - they just lie outside the
    image domain. `clamp_to_domain=False` finds them, at the price of disagreeing with
    CMTK. This is the one semantic knob in the port, so pin both sides of it.
    """
    pts = golden["input"]

    faithful = reg.xform_inv(pts)  # default: clamped, CMTK-faithful
    assert np.isnan(faithful[[0, 4]]).all()

    loose = reg.xform_inv(pts, clamp_to_domain=False)
    assert np.isfinite(loose[0]).all(), "the preimage exists once the domain box is lifted"

    # It is a real preimage: it maps forward onto the target. Verifying that needs
    # `allow_extrapolation=True` -- the preimage is by construction outside the domain box,
    # so the CMTK-faithful forward transform refuses to evaluate there (which is the whole
    # reason CMTK rejects it).
    domain = reg.spacing[0] * (reg.dims[0] - 3)
    assert (loose[0] < 0).any() or (loose[0] > domain).any(), "preimage is out of domain"
    assert np.isnan(reg.xform(loose[0])).all(), "...so the faithful forward declines it"
    np.testing.assert_allclose(
        reg.xform(loose[0], allow_extrapolation=True), pts[0], atol=1e-6
    )

    # Points that solve *inside* the domain are unaffected by the flag.
    np.testing.assert_allclose(faithful[[1, 2, 3]], loose[[1, 2, 3]], atol=1e-9)


def test_initial_guess_is_accepted(reg, golden):
    # Seeding with the answer must not change the answer.
    target = reg.xform(golden["input"][[1, 2, 3]])
    want = reg.xform_inv(target)
    got = reg.xform_inv(target, initial_guess=want)
    np.testing.assert_allclose(got, want, atol=1e-6)


# ---------------------------------------------------------------------------
# Chains
# ---------------------------------------------------------------------------


def test_chain_matches_manual_double_application(reg, cmtk_dir):
    chain = fastcore.CmtkRegistration([cmtk_dir, cmtk_dir])
    assert len(chain) == 2
    pts = np.array([[50.0, 50.0, 50.0], [100.0, 100.0, 20.0]])
    np.testing.assert_allclose(chain.xform(pts), reg.xform(reg.xform(pts)), atol=1e-12)


def test_invert_flag_is_the_inverse_transform(reg):
    """One parse, both directions: `invert` is a property of the traversal, not the object."""
    pts = np.array([[100.0, 100.0, 20.0], [250.0, 150.0, 60.0]])
    np.testing.assert_allclose(reg.xform(reg.xform(pts), invert=True), pts, atol=1e-4)


def test_invert_is_not_the_same_knob_as_xform_inv(cmtk_dir, tiny_dir):
    """`invert` flips each hop in place; `xform_inv` also *reverses* the chain.

    For a single hop they agree. For a chain they must not - and only `invert` can express a
    mixed-direction traversal, which is exactly what a bridging graph hands you. Affine-only,
    so no point can fall out of a domain and make the comparison vacuous through NaN.
    """
    chain = fastcore.CmtkRegistration([cmtk_dir, tiny_dir])
    pts = np.array([[100.0, 100.0, 20.0]])

    flipped = chain.xform(pts, transform="affine", invert=True)
    reversed_ = chain.xform_inv(pts, transform="affine")
    assert np.isfinite(flipped).all() and np.isfinite(reversed_).all()
    assert not np.allclose(flipped, reversed_, atol=1e-6)

    # hop 0 forwards, hop 1 backwards - no whole-chain spelling exists for this
    mixed = chain.xform(pts, transform="affine", invert=[False, True])
    assert np.isfinite(mixed).all()
    assert not np.allclose(mixed, flipped, atol=1e-6)


def test_chain_inverse_reverses_order(reg, cmtk_dir):
    chain = fastcore.CmtkRegistration([cmtk_dir, cmtk_dir])
    pts = np.array([[100.0, 100.0, 20.0], [250.0, 150.0, 60.0]])
    np.testing.assert_allclose(chain.xform_inv(chain.xform(pts)), pts, atol=1e-4)


def test_invert_length_mismatch_raises(reg):
    with pytest.raises(ValueError, match="one flag per registration"):
        reg.xform(np.zeros((2, 3)), invert=[True, False])


# ---------------------------------------------------------------------------
# Loading & introspection
# ---------------------------------------------------------------------------


def test_gz_and_plain_both_load(cmtk_dir, tiny_dir):
    # The bundled JFRC2 registration is gzipped; the tiny fixture is plain text. Gzip is
    # sniffed from magic bytes, not the file name.
    gz = fastcore.CmtkRegistration(cmtk_dir)
    plain = fastcore.CmtkRegistration(tiny_dir)
    np.testing.assert_array_equal(gz.dims, [[59, 27, 11]])
    np.testing.assert_array_equal(plain.dims, [[5, 5, 5]])
    assert gz.version == ["1.1"]
    assert plain.version == ["2.4"]


def test_introspection(reg):
    assert reg.has_spline == [True]
    assert reg.affine.shape == (4, 4)
    np.testing.assert_allclose(reg.affine[3], [0, 0, 0, 1])
    np.testing.assert_allclose(reg.spacing[0], [11.3642, 13.2453, 16.8741], atol=1e-3)
    assert "CmtkRegistration" in repr(reg)


def test_load_cmtk_registration_factory(cmtk_dir):
    reg = fastcore.load_cmtk_registration(cmtk_dir)
    assert isinstance(reg, fastcore.CmtkRegistration)
    assert len(reg) == 1


def test_pickle_roundtrip(reg):
    # navis/joblib fan out over processes, so the object must survive a pickle. It re-loads
    # from disk rather than shipping ~420 KB of coefficients through the pipe.
    pts = np.array([[50.0, 50.0, 50.0], [250.0, 150.0, 60.0]])
    revived = pickle.loads(pickle.dumps(reg))
    np.testing.assert_allclose(revived.xform(pts), reg.xform(pts))


# ---------------------------------------------------------------------------
# Parallelism
# ---------------------------------------------------------------------------


def test_n_cores_does_not_change_the_answer(reg, golden):
    # Each point is solved independently, so the result must be bit-identical regardless of
    # how the work was split.
    one = reg.xform_inv(golden["input"], n_cores=1)
    many = reg.xform_inv(golden["input"])
    np.testing.assert_array_equal(np.isnan(one), np.isnan(many))
    np.testing.assert_array_equal(one[~np.isnan(one)], many[~np.isnan(many)])


def test_large_batch_is_finite(reg):
    rng = np.random.default_rng(0)
    pts = rng.uniform([0, 0, 0], [600, 300, 120], size=(100_000, 3))
    out = reg.xform(pts)
    assert out.shape == (100_000, 3)
    assert np.isfinite(out).all()
